File: TidbitsTrivia/Core/Networking/NightHost.swift
import Foundation
import Observation
import CryptoKit

/// The host side of a Trivia Night (Decision 033) — runs on ANY Apple device
/// (iPhone, iPad, Apple TV). Advertises a room secured by the room-code key,
/// accepts joiners, owns the authoritative roster + standings, and drives the
/// night's pacing on behalf of the human host (who is also a player).
///
/// The host does NOT judge answers — every device runs its own engine over the
/// identical question list and scores itself, reporting its running total. For a
/// friendly living-room game that trust is the right tradeoff (no server, no
/// anti-cheat); the host just aggregates and paces.
///
/// Link-layer-agnostic: all discovery + byte transport lives behind
/// `NightHostTransport` (Bonjour mDNS+TCP by default; Wi-Fi Aware / BLE later).
/// This class owns the protocol, the crypto, and the seat/rejoin model.
@Observable
@MainActor
final class NightHost {
    private(set) var roomCode = ""
    private(set) var isListening = false
    private(set) var lastError: String?
    /// The standings, host included. Seat 0 is always the host.
    private(set) var players: [NightPlayer] = []

    /// How many seats have locked an answer for the current question (host too).
    var answeredCount: Int { players.filter(\.answered).count }
    /// Every connected seat has answered — the host can reveal without waiting.
    var everyoneAnswered: Bool { !players.isEmpty && players.allSatisfy(\.answered) }
    var leaderSeat: Int? {
        guard let top = players.max(by: { $0.score < $1.score }), top.score > 0 else { return nil }
        return top.seat
    }

    static let hostSeat = 0

    private final class Peer {
        let link: any NightPeerLink
        let framer: NightFramer
        var seat: Int?
        init(_ link: any NightPeerLink, key: SymmetricKey) { self.link = link; framer = NightFramer(key: key) }
    }

    private let transport: any NightHostTransport
    private var peers: [String: Peer] = [:]
    private var nextSeat = 1
    /// deviceID → seat, so a reconnecting device resumes its seat + score.
    private var seatByDevice: [String: Int] = [:]
    /// AES-GCM key derived from the room code (v2 app-layer crypto, not TLS).
    private var key = RoomCode.presharedKey(for: "")

    // The live night, retained so a device that joins (or rejoins) mid-game is
    // caught all the way up: the full question list, where we are, and whether
    // the current question is already revealed.
    private var activePlan: NightPlan?
    private var activeQuestions: [Question]?
    private var currentIndex = -1
    private var revealed = false

    init(transport: any NightHostTransport = BonjourHostTransport()) {
        self.transport = transport
    }

    // MARK: Lifecycle

    func start(hostName: String, code: String? = nil) {
        stop()
        roomCode = code ?? RoomCode.generate()
        key = RoomCode.presharedKey(for: roomCode)
        let name = hostName.trimmingCharacters(in: .whitespacesAndNewlines)
        players = [NightPlayer(seat: Self.hostSeat, name: name.isEmpty ? "Host" : name, isHost: true)]
        transport.start(
            roomCode: roomCode,
            onPeer: { [weak self] link in self?.accept(link) },
            onFrame: { [weak self] link, bytes in self?.ingest(bytes, from: link) },
            onDrop: { [weak self] link in
                self?.peers.removeValue(forKey: link.id)
                // Keep the player + score in the roster so a dropped device can
                // REJOIN (same deviceID → same seat) and resume.
            },
            onState: { [weak self] state in
                switch state {
                case .ready:         self?.isListening = true
                case .failed(let e): self?.lastError = e; self?.isListening = false
                case .stopped:       self?.isListening = false
                }
            })
    }

    func stop() {
        transport.stop()
        peers.removeAll(); players.removeAll(); seatByDevice.removeAll()
        isListening = false; nextSeat = 1
        activePlan = nil; activeQuestions = nil; currentIndex = -1; revealed = false
    }

    // MARK: Pacing (called by LiveNight on the host's behalf)

    /// Ship the whole night to everyone, once, at game start.
    func broadcastNight(plan: NightPlan, questions: [Question]) {
        activePlan = plan; activeQuestions = questions
        var m = NightMessage(.night); m.plan = plan
        m.questionIds = questions.map(\.id); m.questions = questions.map { $0.toWire() }
        broadcast(m)
    }

    /// Move everyone to a question. Clears the per-question "answered" flags so the
    /// host's "k of n answered" readout restarts.
    func broadcastBegin(index: Int) {
        currentIndex = index; revealed = false
        for i in players.indices { players[i].answered = false }
        var m = NightMessage(.begin); m.questionIndex = index
        broadcast(m); broadcastRoster()
    }

    /// Reveal the current question's answer on every device at once.
    func broadcastReveal(index: Int) {
        revealed = true
        var m = NightMessage(.reveal); m.questionIndex = index
        broadcast(m)
    }

    /// End the night — everyone shows the final standings.
    func broadcastFinished() {
        broadcast(NightMessage(.finished))
    }

    /// The host (seat 0) locked its own answer — fold it into the standings.
    func setHostAnswered(score: Int, correct: Bool) {
        if let i = players.firstIndex(where: { $0.seat == Self.hostSeat }) {
            players[i].score = score
            players[i].answered = true
        }
        broadcastRoster()
    }

    func name(forSeat seat: Int) -> String {
        players.first { $0.seat == seat }?.name ?? "Player \(seat)"
    }

    // MARK: Connections

    private func accept(_ link: any NightPeerLink) {
        peers[link.id] = Peer(link, key: key)
    }

    private func ingest(_ bytes: Data, from link: any NightPeerLink) {
        guard let peer = peers[link.id] else { return }
        for msg in peer.framer.ingest(bytes) { handle(msg, from: peer) }
    }

    // MARK: Message handling

    private func handle(_ msg: NightMessage, from peer: Peer) {
        switch msg.kind {
        case .join:
            let raw = (msg.displayName ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            let device = msg.deviceID
            // Rejoin by DEVICE first (resume seat + score), else by name, else fresh.
            let resumed: Int? = device.flatMap { seatByDevice[$0] }
                ?? (raw.isEmpty ? nil : players.first { $0.seat != Self.hostSeat && $0.name.caseInsensitiveCompare(raw) == .orderedSame }?.seat)
            let seat: Int
            if let resumed, players.contains(where: { $0.seat == resumed }) {
                seat = resumed
                if !raw.isEmpty, let i = players.firstIndex(where: { $0.seat == seat }) { players[i].name = raw }
            } else {
                seat = nextSeat; nextSeat += 1
                players.append(NightPlayer(seat: seat, name: raw.isEmpty ? "Player \(seat)" : raw))
            }
            if let device { seatByDevice[device] = seat }
            peer.seat = seat
            var welcome = NightMessage(.welcome)
            welcome.seat = seat; welcome.roomName = "Tidbits \(roomCode)"
            send(welcome, to: peer)
            broadcastRoster()
            replayState(to: peer)   // catch a mid-night (re)join all the way up

        case .answered:
            guard let seat = peer.seat, let i = players.firstIndex(where: { $0.seat == seat }) else { return }
            players[i].score = msg.score ?? players[i].score
            players[i].answered = true
            broadcastRoster()

        case .leave:
            peers.removeValue(forKey: peer.link.id)
            peer.link.close()

        default:
            break
        }
    }

    /// Bring a freshly-(re)joined device up to the live state: the whole night,
    /// the current question, and a reveal if this question is already revealed —
    /// so a late arrival lands right IN the game, not on a waiting screen.
    private func replayState(to peer: Peer) {
        guard let plan = activePlan, let questions = activeQuestions else { return }
        var n = NightMessage(.night); n.plan = plan
        n.questionIds = questions.map(\.id); n.questions = questions.map { $0.toWire() }
        send(n, to: peer)
        if currentIndex >= 0 {
            var b = NightMessage(.begin); b.questionIndex = currentIndex
            send(b, to: peer)
            if revealed {
                var r = NightMessage(.reveal); r.questionIndex = currentIndex
                send(r, to: peer)
            }
        }
    }

    // MARK: Send / broadcast

    private func send(_ m: NightMessage, to peer: Peer) {
        guard let frame = NightTransport.encode(m, key: key) else { return }
        peer.link.send(frame)
    }
    private func broadcast(_ m: NightMessage) {
        guard let frame = NightTransport.encode(m, key: key) else { return }
        for p in peers.values { p.link.send(frame) }
    }
    private func broadcastRoster() {
        var m = NightMessage(.roster); m.players = players
        broadcast(m)
    }
}

File: android/app/src/main/java/com/learningischange/tidbitstrivia/net/NightHost.kt
package com.learningischange.tidbitstrivia.net

import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.learningischange.tidbitstrivia.data.Question

/**
 * Host side of a Trivia Night (Decision 033) — transport-agnostic. Owns the
 * authoritative roster + standings and paces the night on the human host's
 * behalf (the host is also a player, seat 0). It does NOT judge answers: every
 * device runs its own engine over the identical id-based question list and
 * reports its running total; the host aggregates and paces.
 *
 * Mirrors the Apple `NightHost`. The link-layer is injected as a
 * [NightHostTransport] (mDNS+TCP today; Wi-Fi Aware / BLE later).
 *
 * NOT yet two-device-verified — a hardware pairing test is the gate.
 */
class NightHost(private val transport: NightHostTransport) {
    var roomCode by mutableStateOf(""); private set
    var isListening by mutableStateOf(false); private set
    var lastError by mutableStateOf<String?>(null); private set
    /** Standings, host included. Seat 0 is the host. */
    val players = mutableStateListOf<NightPlayer>()

    val answeredCount: Int get() = players.count { it.answered }
    val everyoneAnswered: Boolean get() = players.isNotEmpty() && players.all { it.answered }
    val leaderSeat: Int? get() = players.maxByOrNull { it.score }?.takeIf { it.score > 0 }?.seat

    private val main = Handler(Looper.getMainLooper())
    private var key = ByteArray(0)

    private class PeerState(val peer: NightPeer, val framer: NightFramer) { var seat: Int? = null }
    private val peers = HashMap<String, PeerState>()
    private var nextSeat = 1
    private val seatByDevice = HashMap<String, Int>()

    // Retained so a device that (re)joins mid-game is caught all the way up.
    private var activePlan: NightPlan? = null
    private var activeIds: List<String> = emptyList()
    private var currentIndex = -1
    private var revealed = false

    companion object { const val HOST_SEAT = 0 }

    fun start(hostName: String) {
        stop()
        roomCode = RoomCode.generate()
        key = RoomCode.key(roomCode)
        val name = hostName.trim().ifEmpty { "Host" }
        players.add(NightPlayer(seat = HOST_SEAT, name = name, isHost = true))
        transport.start(
            roomCode = roomCode,
            onPeer = { p -> main.post { peers[p.id] = PeerState(p, NightFramer(key)); isListening = true } },
            onFrame = { p, bytes -> main.post { peers[p.id]?.let { st -> st.framer.ingest(bytes).forEach { handle(it, st) } } } },
            onDrop = { p -> main.post { peers.remove(p.id) } },   // keep the seat+score for rejoin
        )
        isListening = true
    }

    fun stop() {
        transport.stop()
        peers.values.forEach { it.peer.close() }
        peers.clear(); players.clear(); seatByDevice.clear()
        isListening = false; nextSeat = 1
        activePlan = null; activeIds = emptyList(); currentIndex = -1; revealed = false
        lastError = null
    }

    // ---- Pacing (called by LiveNight on the host's behalf) ----

    /** Ship the whole night to everyone, once, at start — id-based. */
    fun broadcastNight(plan: NightPlan, questions: List<Question>) {
        activePlan = plan; activeIds = questions.map { it.id }
        broadcast(NightMessage(NightKind.night, plan = plan, questionIds = activeIds))
    }

    fun broadcastBegin(index: Int) {
        currentIndex = index; revealed = false
        for (i in players.indices) players[i] = players[i].copy(answered = false)
        broadcast(NightMessage(NightKind.begin, questionIndex = index)); broadcastRoster()
    }

    fun broadcastReveal(index: Int) {
        revealed = true
        broadcast(NightMessage(NightKind.reveal, questionIndex = index))
    }

    fun broadcastFinished() = broadcast(NightMessage(NightKind.finished))

    /** The host (seat 0) locked its own answer — fold it into the standings. */
    fun setHostAnswered(score: Int, correct: Boolean) {
        val i = players.indexOfFirst { it.seat == HOST_SEAT }
        if (i >= 0) players[i] = players[i].copy(score = score, answered = true)
        broadcastRoster()
    }

    fun nameForSeat(seat: Int): String = players.firstOrNull { it.seat == seat }?.name ?: "Player $seat"

    // ---- Message handling ----

    private fun handle(msg: NightMessage, st: PeerState) {
        when (msg.kind) {
            NightKind.join -> {
                val raw = (msg.displayName ?: "").trim()
                val device = msg.deviceID
                val resumed = device?.let { seatByDevice[it] }
                    ?: raw.takeIf { it.isNotEmpty() }?.let { r -> players.firstOrNull { it.seat != HOST_SEAT && it.name.equals(r, true) }?.seat }
                val seat: Int
                if (resumed != null && players.any { it.seat == resumed }) {
                    seat = resumed
                    if (raw.isNotEmpty()) { val i = players.indexOfFirst { it.seat == seat }; if (i >= 0) players[i] = players[i].copy(name = raw) }
                } else {
                    seat = nextSeat++; players.add(NightPlayer(seat = seat, name = raw.ifEmpty { "Player $seat" }))
                }
                if (device != null) seatByDevice[device] = seat
                st.seat = seat
                send(st.peer, NightMessage(NightKind.welcome, seat = seat, roomName = "Tidbits $roomCode"))
                broadcastRoster()
                replayState(st.peer)
            }
            NightKind.answered -> {
                val seat = st.seat ?: return
                val i = players.indexOfFirst { it.seat == seat }
                if (i >= 0) players[i] = players[i].copy(score = msg.score ?: players[i].score, answered = true)
                broadcastRoster()
            }
            NightKind.leave -> { peers.remove(st.peer.id); st.peer.close() }
            else -> {}
        }
    }

    /** Bring a freshly-(re)joined device up to the live state. */
    private fun replayState(peer: NightPeer) {
        val plan = activePlan ?: return
        send(peer, NightMessage(NightKind.night, plan = plan, questionIds = activeIds))
        if (currentIndex >= 0) {
            send(peer, NightMessage(NightKind.begin, questionIndex = currentIndex))
            if (revealed) send(peer, NightMessage(NightKind.reveal, questionIndex = currentIndex))
        }
    }

    // ---- Broadcast ----

    private fun broadcast(m: NightMessage) { peers.values.forEach { send(it.peer, m) } }
    private fun broadcastRoster() = broadcast(NightMessage(NightKind.roster, players = players.toList()))
    private fun send(peer: NightPeer, m: NightMessage) { NightWire.encode(m, key)?.let { peer.send(it) } }
}
